Review edits to Device Management for Mbed OS (#1046)
ARMmbed/Handbook: docs/pelion_ready/device-management-config.md

Device Management configuration has five distinct areas:

> - Connectivity - the transport type for the device to connect to the Device Management service.

Reviewer: : instead of - ?

Author: Already one : above, so I decided to leave this as is.

- Storage - the storage type and writing used for both the credentials and the firmware storage.

> - Flash geometry - the device flash "sandbox" for bootloader, application header, application image and two Software One Time Programming (SOTP) regions.

Reviewer: One-Time (capitalization?)

Author: We have this without in our glossary. However, the glossary may be wrong.

Reviewer: @AnotherButler, this one's for you.

- SOTP - the address and size of the SOTP regions in flash used to store the device special key that decrypts the credentials storage.
- Bootloader - the bootloader image, application and header offset.

Except for connectivity, the majority of the configuration is shared between the application and bootloader, which ensures the bootloader can correctly find, verify, authorize and apply an update to the device application.

> For more details about bootloaders and firmware update, read the following documents:

Reviewer: ```suggestion
For more information on bootloaders and firmware update, see:
```

Reviewer 2: When we send people elsewhere, we should use "please".

- [Bootloader configuration in Mbed OS](../reference/bootloader-configuration.html).
- [Mbed bootloader for Device Management](https://github.com/ARMmbed/mbed-bootloader), or the short introduction [in the Device Management documentation](https://www.pelion.com/docs/device-management/current/updating-firmware/bootloaders.html).
- [Updating devices with Mbed CLI](../tools/cli-update.html).

To speed up this process, you can copy the configuration from the [application example](https://github.com/ARMmbed/pelion-ready-example/blob/master/mbed_app.json) as the basis for your application configuration.

## Application configuration

Edit the `mbed_app.json` file, and create a new entry under `target_overrides` with the target name for your device:

- **Connectivity** - Specify the default connectivity type for your target. It is essential with targets that lack default connectivity set in `targets.json` or for targets that support multiple connectivity options. For example:

   ```
            "target.network-default-interface-type" : "ETHERNET",
   ```

   The possible options are `ETHERNET`, `WIFI` and `CELLULAR`.

>    Depending on connectivity type, you may have to specify more options. For an example, see the defined `CELLULAR` targets in `mbed_app.json`.

Reviewer 2: Please see earlier comment about "please".

Author: Not willing to be that polite.

- **Storage** - Specify the storage block device type, which dynamically adds the block device driver you specified at compile time. For example:

   ```
            "target.components_add" : ["SD"],
   ```

>    Valid options are `SD`, `SPIF`, `QSPIF` and `FLASHIAP` (not recommended). For more available options, see the [block device components](https://github.com/ARMmbed/mbed-os/tree/master/components/storage/blockdevice).

Reviewer 2: Please see earlier comment about "please".

Author: Still not that polite.

>    You also have to specify the block device pin configuration, which may vary from one block device type to another. Here's an example for `SD`:

Reviewer: ```suggestion
   You must also specify the block device pin configuration, which may vary from one block device type to another. Here's an example for `SD`:
```

            "device-management.sotp-num-sections" : 2
   ```

   - `*-address` defines the start of the flash sector.
   - `*-size` defines the actual sector size. 
   - `sotp-num-sections` should always be set to `2`.

> At this point, we recommend that you run the "connect" test suite, which verifies that the device can successfully bootstrap, register and send and receive data from Device Management with the provided configuration.

Reviewer 2: This "that" is unnecessary.

> If you already configured your Device Management API key and initialized your credentials as described in [Adding Device Management connectivity to your application](../mbed-os-pelion/device-management-for-mbed-os.html#adding-device-management-connectivity-to-your-application), you can compile the "connect" tests as follows:

Reviewer 3: If this is in the same page, you don't need the book and page names in the link. Just the #header

Reviewer 3: You know I remove "as follows" wherever I see it, right?

Author: You are absolutely right. I already forgot it was on the same page.

Author: Then again, I did check that this is not actually on the same page. It is in a different doc, therefore with the relative link.

Reviewer 3: Sorry, my bad

Author: - Fair enough. I'll try to remember to stop using "as follows". 
- I remove "please", whenever I see it. We don't need to be that polite in the docs.

Reviewer 3: We're a British company. We say "please" and make our PRs queue properly.

Author: We are not a British company. We are Japanese, remember.

Reviewer: 🤣

$ mbed test -t <TOOLCHAIN> -m <BOARD> -n simple*dev*connect --run -v
```

## Bootloader configuration

> When the "connect" tests have passed, you can enable the firmware update feature by adding a bootloader to your application.

Reviewer: Are there any circumstances under which someone wouldn't do this? Should it just be an imperative 'enable' minus 'you can'?

> When the "connect" tests have completed successfully, you can enable the firmware update feature by adding a bootloader to your application.

Reviewer 3: Completing and passing tests are not the same thing

Author: Right again. Needs to be passed.

1. Import the official [mbed-bootloader](https://github.com/ARMmbed/mbed-bootloader/) repository as a new application:

    `mbed import mbed-bootloader`
  
    Alternatively, you can import the [mbed-bootloader-extended](https://github.com/ARMmbed/mbed-bootloader-extended/) repository that builds on top of `mbed-bootloader` and extends the support for file systems and storage drivers:
  
    `mbed import mbed-bootloader-extended`.

> 1. In the imported bootloader application, edit the configuration, for example `mbed-bootloader-extended/mbed_app.json`. Add a new target entry and specify:

Reviewer 2: "entry, and"

Author: Academic comma dilemma. I'd rather not have that.

Reviewer 3: I don't think it's academic; +1 for adding it.

   ```

<span class="notes">**Note:** `mbed-bootloader` is primarily optimized for `GCC_ARM`, so you may want to compile it with that toolchain.

> Before adding the bootloader to your application, you need to compile and flash the bootloader and then connect over the virtual com port to ensure the bootloader is running correctly. You can ignore errors related to checksum verification or failure to jump to application - these are expected at this stage.</span>

Reviewer: 'You must', or just 'compile'.

> ## Adding the bootloader to your application

Reviewer 2: Do these steps need to happen in order? If so, is there a way to indicate that without numbering the headings?

Author: Numbering in headings is a bit questionable.

Reviewer: Heading levels, perhaps, but less noticeable in the rendered doc than the raw markdown.

Reviewer 3: Manual numbering goes wrong very quickly. I've asked for automated numbering to be added to the CSS, but we don't have it yet.
